dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,152 @@
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AzureCosmosDBMongoDBVectorStoreRecordMapper{TRecord}"/> class.
+    /// </summary>
+    /// <param name="keyProperty">A property info object that points at the key property for the current model, allowing easy reading and writing of this property.</param>
+    /// <param name="dataProperties">A list of property info objects that point at the data properties in the current model, and allows easy reading and writing of these properties.</param>
+    /// <param name="vectorProperties">A list of property info objects that point at the vector properties in the current model, and allows easy reading and writing of these properties.</param>
+    /// <param name="storagePropertyNames">A dictionary that maps from a property name to the configured name that should be used when storing it.</param>
+    public AzureCosmosDBMongoDBVectorStoreRecordMapper(PropertyInfo keyProperty, List<PropertyInfo> dataProperties, List<PropertyInfo> vectorProperties, Dictionary<string, string> storagePropertyNames)
+    {
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes([keyProperty], s_supportedKeyTypes, "Key");
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(dataProperties, s_supportedDataTypes, "Data", supportEnumerable: true);
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(vectorProperties, s_supportedVectorTypes, "Vector");
+
+        this._storagePropertyNames = storagePropertyNames;
+
+        // Use Mongo reserved key property name as storage key property name
+        this._storagePropertyNames[keyProperty.Name] = AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName;
+
+        this._reversedStoragePropertyNames = ReverseDictionary(storagePropertyNames);
+    }
+
+    public BsonDocument MapFromDataToStorageModel(TRecord dataModel)
+        => MapDocument(dataModel.ToBsonDocument(), this._storagePropertyNames);
+
+    public TRecord MapFromStorageToDataModel(BsonDocument storageModel, StorageToDataModelMapperOptions options)
+        => BsonSerializer.Deserialize<TRecord>(MapDocument(storageModel, this._reversedStoragePropertyNames));
+
+    #region private
+
+    private static BsonDocument MapDocument(BsonDocument document, Dictionary<string, string> propertyMappings)
+    {
+        var newDocument = new BsonDocument();
+
+        foreach (var element in document)
+        {
+            BsonValue newValue;
+            if (element.Value.IsBsonDocument)
+            {
+                newValue = MapDocument(element.Value.AsBsonDocument, propertyMappings);
+            }
+            else if (element.Value.IsBsonArray)
+            {
+                newValue = MapDocumentInArray(element.Value.AsBsonArray, propertyMappings);
+            }
+            else
+            {
+                newValue = element.Value;
+            }
+
+            if (propertyMappings.TryGetValue(element.Name, out var newName))
```

Comment:
If you can apply attributes to the model that affects the naming of BSON serialization names, similar to JsonPropertyName, the names here may not match the names you got from the StoragePropertyName properties.

Reply:
Yes, so my initial plan was that users will need to provide custom mapper if they want to use BSON serialization names, or otherwise add record definition attributes with `StoragePropertyName` (basically use both attributes at the same time). But if you think it's not the best approach, I'm happy to discuss it further, since this is currently an idea only.

Reply:
I think it may not be obvious that if you decorate your model with BSON attributes that you need a custom mapper. I didn't find it to be that hard to support for JSON, but I haven't looked into this BSON serializer so let's chat about it more and we can see.

Reply:
I've added more logic and tests to verify that following models work as expected:
![image](https://github.com/user-attachments/assets/d3a6f943-f022-41f7-96eb-04ff1aa571fa)

Reply:
There is another test case:
```cs
[BsonElement("bson_hotel_name")]
[VectorStoreRecordData(StoragePropertyName = "storage_hotel_name")]
public string? HotelName { get; set; }
```
Since we said we are not supporting BSON names, in this case I would expect the hotel name to be stored in the database under "storage_hotel_name".  Right?
So when reading from the database, we will get a record back with "storage_hotel_name", which will then be copied to a bson doc, and there it should be copied as "bson_hotel_name", because my assumption is that the bson parser will expect that name to be there.  That's how it works for JSON anyway.

Reply:
Added this test case to unit and integration tests which verifies that `BsonElement` is prioritized.